src/pubsub/src/publisher/worker.rs

```diff
@@ -112,27 +102,21 @@ impl Worker {
                     match msg {
                         Some(ToWorker::Publish(msg)) => {
                             let ordering_key = msg.msg.ordering_key.clone();
-                            let outstanding_publishes = pending_batches.entry(ordering_key).or_insert(OutstandingPublishes::new(&self.topic_name));
-                            outstanding_publishes.pending_batch.push(msg);
-                            if outstanding_publishes.pending_batch.len() as u32 >= message_limit || outstanding_publishes.pending_batch.size() >=  byte_threshold {
-                                outstanding_publishes.pending_batch.flush(self.client.clone(), self.topic_name.clone(), &mut inflight);
-                            }
+                            let outstanding_publishes =
+                                pending_batches
+                                    .entry(ordering_key)
+                                    .or_insert(OutstandingPublishes::new(
+                                        &self.topic_name,
+                                        message_limit,
+                                        byte_threshold,
+                                    ));
+                            outstanding_publishes.push(msg, self.client.clone(), self.topic_name.clone());
                         },
                         Some(ToWorker::Flush(tx)) => {
+                            // TODO(#4012): To guarantee ordering, we should wait for the
+                            // inflight batch to complete so that messages are publish in order.
```

Comment:
This comment feels like it belongs within OutstandingPublish NOT here, since this is no longer responsible for deciding when to send the message.